rsconf/component/devbox.py

```diff
@@ -0,0 +1,113 @@
+            return
+        self.buildt.require_component("docker", "network")
+        jc, _ = self.j2_ctx_init()
+        z = jc.devbox
+        z.setdefault("volumes", ["jupyter", "src"])
+        z.host_d = systemd.unit_run_d(jc, "devbox_" + self.user_name)
+        z.secrets = self._gen_host_and_identity_ssh_keys(jc)
+        for x, d in ("guest", ".ssh"), ("host", "sshd"):
+            z[x] = self._gen_paths(z, z[x + "_d"], d)
+        z.run_u = jc.rsconf_db.run_u
+        # Only additional config for the server is the sshd config.
+        z.run_d = systemd.custom_unit_prepare(
+            self, jc, watch_files=[z.host.ssh_d], run_d=z.host_d
+        )
+        self._network(jc, z)
+
+    def internal_build_write(self):
+        from rsconf import systemd
+
+        if "user_name" not in self:
+            self.append_root_bash(": user instances do all the installs")
+            return
+        jc = self.j2_ctx
+        z = jc.devbox
+        v = [[z.host[v], z.guest[v], "rw"] for v in z.volumes]
+        v.extend(
+            [
+                # SECURITY: no modifications to run_d or ssh_d
+                [z.host.ssh_d, z.guest.ssh_d, "ro"],
+                [z.run_d, z.run_d, "ro"],
+            ],
+        )
+        systemd.docker_unit_enable(
+            self,
+            jc,
+            image=jc.devbox.docker_image,
+            volumes=v,
+            cmd="/usr/sbin/sshd -D -f '{}'".format(z.guest.ssh_d.join("sshd_config")),
+        )
+        self.install_access(mode="700", owner=z.run_u)
+        for d in map(lambda x: x[0], v):
+            self.install_directory(d)
+            if "src" in str(d):
+                r = ""
+                p = d
+                for c in ("biviosoftware", "home-env"):
+                    p = p.join(c)
+                    self.install_directory(p)
+                    r += f"/{c}"
+                self.append_root_bash(
+                    f"rsconf_clone_repo https://github.com{r}.git {p} {jc.rsconf_db.run_u}"
+                )
+        self.install_access(mode="400")
+        self.install_resource(z.host.sshd_config, jc)
+        for k, v in z.secrets.items():
+            self.install_abspath(v, z.host[k])
+
+    def _network(self, jc, z):
+        n = self.buildt.get_component("network")
+        z.ip, _ = n.ip_and_net_for_host(jc.rsconf_db.host)
+        z.ssh_port = jc.devbox.users[self.user_name]
+
+    def _gen_host_and_identity_ssh_keys(self, jc):
+        from rsconf import db
+
+        p = db.random_string()
+        res = super()._gen_host_and_identity_ssh_keys(
+            jc, "devbox/" + self.user_name, visibility="host", password=p
+        )
+        for k in list(res.keys()):
+            if k not in ("host_key_f", "identity_pub_f"):
+                res.pkdel(k)
+        f = db.secret_path(jc, _PASSWD_SECRET_JSON_F, visibility="host")
+        o = pkjson.load_any(f) if f.check() else PKDict()
+        o[self.user_name] = p
```

Comment:
This will change the password every time. You want to do this first so you can get the password for the user or create it. See component.rsconf.host_init.

Reply:
0b2eeca